### ISSUE TITLE
Allow priors to take a param_transform

### ISSUE THREAD

**Issue by Balandat** [cornellius-gp/gpytorch#374]
#371 added support for passing a `param_transform` into many modules to swap out the default exponential transform (i.e. internal parameters in log scale). 

The priors currently still only support a `log_scale` boolean flag, so there is no easy way to put an existing prior on the transformed parameter (without implementing a new prior). We should change this and support passing in a `param_transform` to the prior constructor as well.

**Comment by Balandat** [comment]
This is a good opportunity to also tackle #181 and #339

**Comment by Balandat** [comment]
In line with #181, instead of passing the transform into the prior, we could instead just use what's currently a "derived prior" and register the transform there. The only restriction of that would be that we couldn't have different transforms for parameters in the evaluation vs. the prior. I guess while that's not a typical use case, it would allow to do things like simply put a log-normal prior on a parameter that internally is parameterized using the (inverse) softplus function.

One way to allow for that would be to handle transforms in two places:
1. Always feed the transformed parameter into the prior by doing what #181 suggests, using the construct of the derived_prior (employing `param_transform`)
2. Allow priors themselves to take an optional transform arg, which could be used to transform whatever input it gets. 

That way, if you wanted to put a log-normal prior on the actual lengthscales (which could be transformed via softplus internally), you could use `log` as the transform in a NormalPrior.

@jacobrgardner, @gpleiss any thoughts on this?

**Comment by gpleiss** [comment]
I think I'm fine with either. If I'm understanding what you're suggesting, 2 could be an optional syntactic sugar for 1, right?

**Comment by Balandat** [comment]
Hmm so working on this it seems it would make sense internally to associate the transform with the parameter itself, not the module the parameter is registered with. The constructors for the user-facing api could remain the same, but the transform would be registered with the parameter, similar to what's currently done for the derived prior transforms. 

E.g. for kernels that would look like the following:
```
self.register_parameter(
    name="log_lengthscale",
    parameter=torch.nn.Parameter(torch.zeros(batch_size, 1, lengthscale_num_dims)),
    param_transform=param_transform,
    prior=log_lengthscale_prior,
)

@property
def lengthscale(self):
    if self.has_lengthscale:
        return self._transform_param("log_lengthscale", self.log_lengthscale).clamp(self.eps, 1e5)
    else:
        return None
```

where `_transform_param` is defined on the module level and returns the previously registered transform of the param, or just the param in case there is none registered. 

This way it's easy to re-use the transforms in other places without having to do a lot of special casing. To make this more concrete I'll put up a PR later today.

**Comment by jacobrgardner** [comment]
Yeah, I think registering the transform with the parameter makes sense.
